### DISCUSSION TITLE
how to load wfs features?

### DISCUSSION THREAD

**Discussion by GISlxz** [pelicanmapping/osgearth#2629] [Support Q&A] [answered]
I am trying to load a WFS layer published by GeoServer using CLASS WFSFeatureSource, but the features are not being read correctly. I have checked the source code in WFS.cpp to verify the URL construction and ensure my parameters are correct, but the features are still not loading properly.

Additionally, I tried using .earth to load the WFS layer, and while no loading errors were displayed, the features still do not show up correctly.

    std::string source = "http://localhost:8084/geoserver/wfs?";
    osg::ref_ptr<WFSFeatureSource> wfsSource = new WFSFeatureSource;
    wfsSource->setURL(osgEarth::URI(source));
    wfsSource->setTypeName("countries");
    wfsSource->setOutputFormat("GML2");
    //wfsSource->setOutputFormat("json");
    auto count = wfsSource->getFeatureCount();
    std::cout << count; // result:-1

    osg::ref_ptr<FeatureModelLayer> wfsLayer = new FeatureModelLayer();
    wfsLayer->setName("wfsLayer");
    wfsLayer->setFeatureSource(wfsSource);
    map->addLayer(wfsLayer);

thanks a lot

**Comment by gwaldron** [comment, marked as answer]
I believe the legal options for setOutputFormat are "geojson" and "gml".

Also, WFS does not implement `getFeatureCount()` so it will always return -1. Sometimes there is no way to query that information without actually fetching the data. Most network-based feature sources are like this.

I will be honest with you; you may be the first person to use the WFS layer in a long time. There's a good chance it is very out of date and needs attention. Let us know if you have any further luck.
